### PR TITLE
Build for qnx7

### PR DESCRIPTION
**Pull request by BSolonenko**
This PR is **ready** for review.  
This PR includes implementation for  proposal [Porting SDL to QNX700 and Automotive Grade Linux (AGL) x64 platforms](https://github.com/LuxoftAKutsan/sdl_evolution/blob/port_sdl_on_qnx700_and_agl/proposals/nnnn-sdl_on_qnx700_agl.md)

### Summary
In this PR was added getter and setter for path in qdb sql database, and fixed Build for QNX.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Review by AKalinich-Luxoft — APPROVED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,6 +58,11 @@ option(USE_GOLD_LD "Use gold linker intead of GNU linker" ON)
 option(USE_CCACHE "Turn on ccache usage" ON)
 option(USE_DISTCC "Turn on distributed build_usage" OFF)
 
+if(UNIX AND QNXNTO)
+  set(BUILD_BT_SUPPORT OFF)
```
Comment (LuxoftAKutsan) [inline CMakeLists.txt]: @BSolonenko can we make BUILD_BT_SUPPORT and BUILD_USB_SUPPORT as options?

File: CMakeLists.txt [hunk 2]
```diff
@@ -235,7 +242,12 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
 set(CMAKE_INSTALL_PREFIX ${CMAKE_CURRENT_BINARY_DIR})
 set(ARCHIVE_OUTPUT_DIRECTORY ./bin)
 
-set(CMAKE_CXX_FLAGS "-fPIC -std=gnu++0x -Wall -Werror -Wuninitialized -Wvla")
+set(CMAKE_CXX_FLAGS "-fPIC -std=gnu++11  -Wall -Werror -Wuninitialized -Wvla")
```
Comment (LuxoftAKutsan) [inline CMakeLists.txt]: @BSolonenko can we try to use cmake instead of gcc flags ?

File: CMakeLists.txt [hunk 3]
```diff
@@ -235,7 +242,12 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
...
+
+get_os(OS)
+if(${OS} STREQUAL "qnx")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libstdc++")
```
Comment (LuxoftAKutsan) [inline CMakeLists.txt:252]: @BSolonenko is this option applicable only to QNX? may be we can make it without if

File: src/3rd_party/QNX_Boost.cmake [hunk 4]
```diff
@@ -43,19 +43,21 @@ BOOTSTRAP
...
+set(BOOST_BUILD_COMMAND ./b2 address-model=${ADDRESS_MODEL} cxxflags="-stdlib=libstdc++" linkflags="-stdlib=libstdc++" target-os=qnxnto toolset=gcc-nto${CMAKE_SYSTEM_PROCESSOR} define=__QNXNTO__)
 set(BOOST_INSTALL_COMMAND ${BOOST_BUILD_COMMAND} install > boost_install.log)
 if (${3RD_PARTY_INSTALL_PREFIX} MATCHES "/usr/local")
   set(BOOST_INSTALL_COMMAND sudo  ${BOOST_BUILD_COMMAND} install > boost_install.log)
 endif()
 include(ExternalProject)  
 ExternalProject_Add(
   Boost
-  URL https://dl.bintray.com/boostorg/release/1.66.0/source/boost_1_66_0.tar.gz
+  URL http://dl.bintray.com/boostorg/release/1.66.0/source/boost_1_66_0.tar.gz
```
Comment (LuxoftAKutsan) [inline src/3rd_party/QNX_Boost.cmake]: @BSolonenko what is the reason of not secure pull?

File: src/components/application_manager/include/application_manager/resumption/resumption_data_db.h [hunk 5]
```diff
@@ -34,8 +34,13 @@
 #define SRC_COMPONENTS_APPLICATION_MANAGER_INCLUDE_APPLICATION_MANAGER_RESUMPTION_RESUMPTION_DATA_DB_H_
 #include "application_manager/resumption/resumption_data.h"
 
+#ifdef __QNX__
+#include "utils/qdb_wrapper/sql_database.h"
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/resumption/resumption_data_db.h]: @BSolonenko can we try to resolve this  cross platform issue in cmakes?

File: src/components/media_manager/CMakeLists.txt [hunk 6]
```diff
@@ -67,6 +67,10 @@ set(LIBRARIES
   Utils
 )
 
+if(${OS} STREQUAL "qnx")
+  list(APPEND LIBRARIES socket)
```
Comment (LuxoftAKutsan) [inline src/components/media_manager/CMakeLists.txt:71]: @BSolonenko can we try to use cmake generators? Please check if it compatible with current cmake version

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 7]
```diff
@@ -38,7 +38,11 @@
 
 #include "utils/logger.h"
 #include "utils/date_time.h"
+#if __QNX__
+#include "utils/qdb_wrapper/sql_database.h"
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @BSolonenko please try to resolve this issue with cmakes

File: src/components/utils/CMakeLists.txt [hunk 8]
```diff
@@ -110,8 +110,20 @@ else ()
     add_subdirectory(./src/sqlite_wrapper)
 endif ()
 
-if(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
- list(APPEND LIBRARIES dl pthread ${RTLIB})
+if(UNIX AND NOT QNXNTO)
```
Comment (LuxoftAKutsan) [inline src/components/utils/CMakeLists.txt:113]: @BSolonenko confusing setting libraries in IF, and sources in ELSE. Please do some small refactor to avoid it

File: toolchains/Toolchain-QNX7-QCC-AArch64.cmake [hunk 9]
```diff
@@ -0,0 +1,6 @@
+set(CMAKE_SYSTEM_PROCESSOR aarch64)
```
Comment (LuxoftAKutsan) [inline toolchains/Toolchain-QNX7-QCC-AArch64.cmake]: @BSolonenko this toolchain is redundant

**Review by LuxoftAKutsan — CHANGES_REQUESTED** [review]

File: toolchains/Toolchain-QNX7-QCC-AArch64.cmake (hunk 9 above)
Comment (BSolonenko) [reply toolchains/Toolchain-QNX7-QCC-AArch64.cmake]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: src/components/utils/CMakeLists.txt (hunk 8 above)
Comment (BSolonenko) [reply src/components/utils/CMakeLists.txt:113]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 7 above)
Comment (BSolonenko) [reply src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: src/components/application_manager/include/application_manager/resumption/resumption_data_db.h (hunk 5 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/resumption/resumption_data_db.h]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: src/3rd_party/QNX_Boost.cmake (hunk 4 above)
Comment (BSolonenko) [reply src/3rd_party/QNX_Boost.cmake]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: CMakeLists.txt (hunk 3 above)
Comment (BSolonenko) [reply CMakeLists.txt:252]: @LuxoftAKutsan No, we can`t. This option is applicable only to QNX.

File: CMakeLists.txt (hunk 2 above)
Comment (BSolonenko) [reply CMakeLists.txt]: @LuxoftAKutsan Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: CMakeLists.txt (hunk 1 above)
Comment (BSolonenko) [reply CMakeLists.txt]: @LuxoftAKutsan  Fixed in b2b51c312fc57007af9a6626c0e88bfe2f0b594e

File: src/components/media_manager/CMakeLists.txt (hunk 6 above)
Comment (BSolonenko) [reply src/components/media_manager/CMakeLists.txt:71]: @LuxoftAKutsan  What do you mean?

**Review by SNiukalov — APPROVED** [review]

File: src/components/application_manager/include/application_manager/app_launch/app_launch_data_db.h [hunk 10]
```diff
@@ -36,13 +36,10 @@
...
 #include "utils/macro.h"
-#ifdef __QNX__
-#include "utils/qdb_wrapper/sql_database.h"
-#include "utils/qdb_wrapper/sql_query.h"
-#else  // linux
-#include "utils/sqlite_wrapper/sql_database.h"
-#include "utils/sqlite_wrapper/sql_query.h"
-#endif
+
+#include "sql_database.h"
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/app_launch/app_launch_data_db.h]: @BSolonenko Use this rule: [Order of includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h [hunk 11]
```diff
@@ -91,6 +91,19 @@ class SQLDatabase {
    */
   bool Backup();
 
+	/**
```
Comment (AByzhynar) [inline src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @BSolonenko indents...

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h [hunk 12]
```diff
@@ -91,6 +91,19 @@ class SQLDatabase {
...
+   * Sets path to database
+   * If the database is already opened then need reopen it
```
Comment (AByzhynar) [inline src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @BSolonenko Confusing comment

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h [hunk 13]
```diff
@@ -91,6 +91,19 @@ class SQLDatabase {
...
+   */
+  void set_path(const std::string& path);
+
+   /**
+   * @brief get_path databse location path.
```
Comment (AByzhynar) [inline src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @BSolonenko typo

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h [hunk 14]
```diff
@@ -91,6 +91,19 @@ class SQLDatabase {
...
+   *
```
Comment (AByzhynar) [inline src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @BSolonenko 
Redundant empty line

File: src/components/utils/src/qdb_wrapper/sql_query.cc [hunk 15]
```diff
@@ -103,8 +103,6 @@ SQLQuery::SQLQuery(SQLDatabase* db)
 
 SQLQuery::~SQLQuery() {
   Finalize();
-  db_->Close();
-  delete db_;
```
Comment (AByzhynar) [inline src/components/utils/src/qdb_wrapper/sql_query.cc]: @BSolonenko Why did you delete it?

File: toolchains/Toolchain-QNX7-QCC-Base.cmake [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+file(READ "${QNX_TARGET}/etc/qversion" CMAKE_SYSTEM_VERSION)
+string(REGEX REPLACE "[^0-9.]" "" CMAKE_SYSTEM_VERSION "${CMAKE_SYSTEM_VERSION}")
+if("${CMAKE_SYSTEM_VERSION}" VERSION_LESS "7.0.0")
+    message(FATAL_ERROR "This toolchain file is not supported for QNX ${CMAKE_SYSTEM_VERSION}")
+else()
+    message(STATUS "QNX Version is ${CMAKE_SYSTEM_VERSION}")
+endif()
+
+# Since we are passing while crosscompilation CMAKE_FIND_ROOT_PATH(containing QNX third party dependencies)
```
Comment (AByzhynar) [inline toolchains/Toolchain-QNX7-QCC-Base.cmake]: @BSolonenko Re-phrase it . Make it shorter and grammatically correct

File: toolchains/Toolchain-QNX7-QCC-Base.cmake [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+# as command line argument to cmake we have to take it into account when setting CMAKE_FIND_ROOT_PATH
+# variable below.
+# ${QNX_TARGET}/usr contains platform independent headers.
+# But QNX_TARGET is a root directory containing sysroots for different0 platforms.
```
Comment (AByzhynar) [inline toolchains/Toolchain-QNX7-QCC-Base.cmake]: @BSolonenko typo

File: toolchains/Toolchain-QNX7-QCC-Base.cmake [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+# Therefore to avoid passing overall QNX_TARGET as root path to search just for 'usr' directory
+# ${QNX_TARGET}/usr passed instead.
+set(CMAKE_FIND_ROOT_PATH "${CMAKE_FIND_ROOT_PATH}" "${QNX_TARGET}/${QNX_TARGET_ARCH_DIR}" "${QNX_TARGET}/usr")
+# Since CMAKE_TOOLCHAIN_FILE behaviour does NOT suppose any 'append' operations like setting
+# CMAKE_FIND_ROOT_PATH above and CMAKE_TOOLCHAIN_FILE must provide the same environment independently
+# of how often and how many times it calls remove duplicates in CMAKE_FIND_ROOT_PATH.
```
Comment (AByzhynar) [inline toolchains/Toolchain-QNX7-QCC-Base.cmake]: @BSolonenko Did not get the meaning of this comment

**Review by AByzhynar — CHANGES_REQUESTED** [review]

**Comment by BSolonenko** [comment]
@LuxoftAKutsan @AByzhynar  In a3a73a446d1a8fe30f7b98dde719b35719b6d731 fixed all you comments.

File: toolchains/Toolchain-QNX7-QCC-Base.cmake [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+set(CMAKE_CXX_COMPILER "${QNX_HOST}/usr/bin/qcc${HOST_EXECUTABLE_SUFFIX}"                                  CACHE PATH "QNX g++ Program")
+set(CMAKE_AR           "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ar${HOST_EXECUTABLE_SUFFIX}"      CACHE PATH "QNX ar Program")
+set(CMAKE_RANLIB       "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ranlib${HOST_EXECUTABLE_SUFFIX}"  CACHE PATH "QNX ranlib Program")
+set(CMAKE_NM           "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-nm${HOST_EXECUTABLE_SUFFIX}"      CACHE PATH "QNX nm Program")
+set(CMAKE_OBJCOPY      "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-objcopy${HOST_EXECUTABLE_SUFFIX}" CACHE PATH "QNX objcopy Program")
+set(CMAKE_OBJDUMP      "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-objdump${HOST_EXECUTABLE_SUFFIX}" CACHE PATH "QNX objdump Program")
+set(CMAKE_LINKER       "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ld${HOST_EXECUTABLE_SUFFIX}"      CACHE PATH "QNX Linker Program")
+set(CMAKE_STRIP        "${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-strip${HOST_EXECUTABLE_SUFFIX}"   CACHE PATH "QNX Strip Program")
+
+# Will only add -V<target> if compiler name is "qcc"
```
Comment (AByzhynar) [inline toolchains/Toolchain-QNX7-QCC-Base.cmake]: @BSolonenko Unclear comment

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: toolchains/Toolchain-QNX7-QCC-Base.cmake (hunk 19 above)
Comment (BSolonenko) [reply toolchains/Toolchain-QNX7-QCC-Base.cmake]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/pull/11/commits/7e960d64e295cc6a681b8bb5732521173ef4a003

File: src/components/utils/src/qdb_wrapper/sql_query.cc (hunk 15 above)
Comment (BSolonenko) [reply src/components/utils/src/qdb_wrapper/sql_query.cc]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/pull/11/commits/7e960d64e295cc6a681b8bb5732521173ef4a003

**Review by LuxoftAKutsan — APPROVED** [review]

File: toolchains/Toolchain-QNX7-QCC-Base.cmake (hunk 18 above)
Comment (BSolonenko) [reply toolchains/Toolchain-QNX7-QCC-Base.cmake]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: toolchains/Toolchain-QNX7-QCC-Base.cmake (hunk 17 above)
Comment (BSolonenko) [reply toolchains/Toolchain-QNX7-QCC-Base.cmake]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: toolchains/Toolchain-QNX7-QCC-Base.cmake (hunk 16 above)
Comment (BSolonenko) [reply toolchains/Toolchain-QNX7-QCC-Base.cmake]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h (hunk 14 above)
Comment (BSolonenko) [reply src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h (hunk 13 above)
Comment (BSolonenko) [reply src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h (hunk 12 above)
Comment (BSolonenko) [reply src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: src/components/utils/include/utils/qdb_wrapper/sql_database.h (hunk 11 above)
Comment (BSolonenko) [reply src/components/utils/include/utils/qdb_wrapper/sql_database.h]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

File: src/components/application_manager/include/application_manager/app_launch/app_launch_data_db.h (hunk 10 above)
Comment (BSolonenko) [reply src/components/application_manager/include/application_manager/app_launch/app_launch_data_db.h]: @AByzhynar fixed in https://github.com/Ford-Luxoft/sdl_core/commit/a3a73a446d1a8fe30f7b98dde719b35719b6d731

**Review by AByzhynar — APPROVED** [review]
